File: backend/app/core/key_limiter.py
import threading
import time

class KeyLimiter:
    def __init__(self):
        self._limits = {}
        self._lock = threading.Lock()
        self.cooldown_period = 60 # seconds

    def is_rate_limited(self, key):
        with self._lock:
            now = time.time()
            limit_data = self._limits.get(key)
            if limit_data and now < limit_data["cooldown"]:
                return True
            return False

    def trigger_cooldown(self, key):
        with self._lock:
            self._limits[key] = {"cooldown": time.time() + self.cooldown_period}

    def record(self, key):
        # Could implement more complex usage tracking here
        pass

File: src/backend/core/keyLimiter.ts
import { Mutex } from 'async-mutex';

export class KeyLimiter {
  private limits: Map<string, number>;
  private mutex: Mutex;
  private cooldownPeriod: number;

  constructor() {
    this.limits = new Map();
    this.mutex = new Mutex();
    this.cooldownPeriod = 60000; // ms
  }

  async isRateLimited(key: string): Promise<boolean> {
    const release = await this.mutex.acquire();
    try {
      const now = Date.now();
      const cooldownUntil = this.limits.get(key);
      if (cooldownUntil && now < cooldownUntil) {
        return true;
      }
      return false;
    } finally {
      release();
    }
  }

  async triggerCooldown(key: string): Promise<void> {
    const release = await this.mutex.acquire();
    try {
      this.limits.set(key, Date.now() + this.cooldownPeriod);
    } finally {
      release();
    }
  }

  record(key: string): void {
    // Tracking usage
  }
}
